File: src/widget.ts
// Copyright (c) Ognyan Moore
// Distributed under the terms of the Modified BSD License.

import {
  DOMWidgetModel,
  DOMWidgetView,
  ISerializers,
} from '@jupyter-widgets/base';

import { MODULE_NAME, MODULE_VERSION } from './version';

// Import the CSS
import '../css/widget.css';

export class EptiumModel extends DOMWidgetModel {
  defaults() {
    return {
      ...super.defaults(),
      _model_name: EptiumModel.model_name,
      _model_module: EptiumModel.model_module,
      _model_module_version: EptiumModel.model_module_version,
      _view_name: EptiumModel.view_name,
      _view_module: EptiumModel.view_module,
      _view_module_version: EptiumModel.view_module_version,
      src: 'https://viewer.copc.io',
    };
  }

  static serializers: ISerializers = {
    ...DOMWidgetModel.serializers,
    // Add any extra serializers here
  };

  static model_name = 'EptiumModel';
  static model_module = MODULE_NAME;
  static model_module_version = MODULE_VERSION;
  static view_name = 'EptiumView'; // Set to null if no view
  static view_module = MODULE_NAME; // Set to null if no view
  static view_module_version = MODULE_VERSION;
}

export class EptiumView extends DOMWidgetView {
  private _iframe: HTMLIFrameElement;

  render() {
    this._iframe = document.createElement('iframe');
    this._iframe.width = '100%';
    this._iframe.height = '400';
    this._iframe.style.boxSizing = 'border-box';
    this._iframe.style.border = '4px solid red';

    // this.el is the DOM element associated with the view
    this.el.appendChild(this._iframe);
    this._onSrcChanged();

    // Python -> TypeScript update
    this.model.on('change:src', this._onSrcChanged, this);
  }

  private _onSrcChanged() {
    const url = this.model.get('src');
    this.model.widget_manager.resolveUrl(url).then((resolvedUrl) => {
      this._iframe.src = resolvedUrl;
    });
  }
}

File: eptium/widget.py
#!/usr/bin/env python
# coding: utf-8

# Copyright (c) Ognyan Moore.
# Distributed under the terms of the Modified BSD License.

"""
TODO: Add module docstring
"""

__all__ = ["Eptium"]

import os
import pathlib
from urllib.parse import urlparse, urlencode

from ipywidgets import DOMWidget, ValueWidget, register
from traitlets import Unicode, validate, TraitError

import requests
from jupyter_server import serverapp
from ._frontend import module_name, module_version

@register
class Eptium(DOMWidget, ValueWidget):
    _model_name = Unicode('EptiumModel').tag(sync=True)
    _model_module = Unicode(module_name).tag(sync=True)
    _model_module_version = Unicode(module_version).tag(sync=True)

    _view_name = Unicode('EptiumView').tag(sync=True)
    _view_module = Unicode(module_name).tag(sync=True)
    _view_module_version = Unicode(module_version).tag(sync=True)

    # read-write attributes
    src = Unicode('https://viewer.copc.io').tag(sync=True)
    height = Unicode('400px').tag(sync=True)

    def __init__(
        self,
        src="https://viewer.copc.io",
    ):
        super().__init__()
        self.src = src
        self.height

    def set_height(self, value: int | str):
        """Set the height of the Eptium View

        Parameters
        ----------
        value : int | str
            Accepted values are used to set the ``height`` attribute
            of an iframe. 

        Examples
        --------
        >>> widget.set_height('200px')
        """
        if isinstance(value, int):
            value = str(value)
        self.height = value

    def render(self, path: str | pathlib.Path):
        """Method to call to generate the Eptium View

        Parameters
        ----------
        path : str | pathlib.Path
            Path to the asset that Eptium should display. Acceptable
            values include local file paths, or URLs to 
        """
        if isinstance(path, pathlib.Path):
            # not using os.fsdecode since we want forward-slashes
            # even on windows
            path = path.as_posix()
        parsed_url = urlparse(path)

        if parsed_url.scheme in ("https", "http"):
            path = f"https://viewer.copc.io/?q={path}"
        else:
            # we're dealing with a local file and need to
            # construct a remote accessible URL
            # TODO: maybe the first server isn't the one we want?
            # should check for a valid URL for all running servers
            server = next(serverapp.list_running_servers())
            cookies = {}

            # ensure token is good
            r = requests.post(
                url=f"{server['url']}api/contents",
                headers={'Authorization': f"token {server['token']}"},
                cookies=cookies
            )
            r.raise_for_status()

            # get the _xsrf cookie
            r = requests.get(
                url=f"{server['url']}lab/tree",
                cookies=r.cookies
            )

            # order matters!
            params = urlencode({
                'token': server['token'],
                '_xsrf': r.cookies['_xsrf']

            })
            
            path = f"https://viewer.copc.io/?q={server['url']}files/{path}?{params}"

        self.src = path
